### ISSUE TITLE
oe_backtrace causes crash in ReleaseWithDebug and Debug

### ISSUE THREAD

**Issue by anakrish** [Microsoft/openenclave#730]
We noticed strange crashes when running the remote_attestation sample under the following scenarios:
1. OE built in debug mode with GCC. Remote attestation sample built in release mode with clang.
2. OE built in ReleaseWithDebug mode with clang. Remote attestation sample built with releasewithdebug mode with clang.
And many more.

The root cause turns out to be that 
1. oe_debug_malloc is turned on when internally uses oe_backtrace for allocation.
2. oe_backtrace uses __builtin_return_address intrinsic which generates callstack traversal code as below:

__builtin_return_address(4) generates 
  movq 0(%rbp), %rax
  movq (%rax), %rax
  movq (%rax), %rax
  movq (%rax), %rax
  movq 8(%rax), %rax

The code to walk the stack does not handle null-pointers or invalid pointers.
With clang, some of the pointers seem to be null, possible because it is not preserving the stack frame.

To summarize:
1. oe_debug_malloc seemed to be turned on for ReleaseWithDebug builds too.
2. oe_backtrace seems to be unstable with clang (and maybe gcc too).

@CodeMonkeyLeet @mikbras @soccerGB @acl33 @letmaik

**Comment by anakrish** [comment]
Fixed by #746
